crates/typed-store/src/rocks/mod.rs

```diff
@@ -94,11 +113,59 @@ pub struct DBMap<K, V> {
     db_metrics: Arc<DBMetrics>,
     read_sample_interval: SamplingInterval,
     write_sample_interval: SamplingInterval,
+    metric_reporter: Arc<Mutex<Option<JoinHandle<()>>>>,
+    exit_signal: Arc<AtomicBool>,
 }
 
 unsafe impl<K: Send, V: Send> Send for DBMap<K, V> {}
 
 impl<K, V> DBMap<K, V> {
+    pub(crate) fn new(
+        db: Arc<rocksdb::DBWithThreadMode<MultiThreaded>>,
+        opt_cf: &str,
+        db_metrics: Arc<DBMetrics>,
+    ) -> Self {
+        let exit = Arc::new(AtomicBool::new(false));
+        let exit_cloned = exit.clone();
+        let db_cloned = db.clone();
+        let db_metrics_cloned = db_metrics.clone();
+        let cf = opt_cf.to_string();
+        let metric_reporter = std::thread::spawn(move || {
+            while !exit.load(Ordering::SeqCst) {
+                Self::report_metrics(&db, &cf, &db_metrics);
+                sleep(Duration::from_millis(CF_METRICS_REPORT_PERIOD_MILLIS));
+            }
+            info!(
+                "Stopping the column family metric reporting task for DBMap: {}",
+                &cf
+            );
+        });
+        DBMap {
+            rocksdb: db_cloned,
+            _phantom: PhantomData,
+            cf: opt_cf.to_string(),
+            db_metrics: db_metrics_cloned,
+            read_sample_interval: SamplingInterval::default(),
+            write_sample_interval: SamplingInterval::default(),
+            metric_reporter: Arc::new(Mutex::new(Some(metric_reporter))),
+            exit_signal: exit_cloned,
+        }
+    }
+
+    pub async fn stop(&self) -> Result<(), TypedStoreError> {
```

Comment:
This is not called anywhere?

Tbh, I think this went off the initially correct corse of stopping the reporting task when the last cloned instance of DBMap is dropped, and now we ended up having this `async stop` method for which it is not clear who is going to call it.

I think it is also ok not to join on the reporting thread to finish, as I suspect that making this join work correctly can be a bit tricky

Anyway, what about following proposal:

(1) Make `exit_signal: Arc<()>` 
(2) Keep element of this `Arc<()>` in DBMap
(3) In the reporting thread have `while Arc::strong_count(exit_signal) > 1`

This way reporting thread stops when DBMap is dropped, no need to dedicated `stop` method

Reply:
The idea is to have stop() be called in the exit routine of wherever DBMap is being used (sui, etc). The motivation is to cleanup any threads before exiting the program. The problem I see with Arc::strong_count is it does things the other way and does not guarantee a proper waiting/shutdown.

Reply:
The problem with this is there is no guarantee of it ever cleaning up the reporting thread. If the sleep within the loop was any longer, the main thread would drop the Arc<> but the reporting thread may never get to checking it and the program terminates with lingering threads. Having a stop() method is a guarantee that the user of this library gets that by invoking they will get the metric reported one last time before the thread terminates.